flink/v1.15/flink/src/main/java/org/apache/iceberg/flink/sink/v2/FlinkSink.java

```diff
@@ -0,0 +1,603 @@
+                   @Nullable String uidPrefix,
+                   ReadableConfig readableConfig,
+                   Map<String, String> writeOptions,
+                   Map<String, String> snapshotProperties) {
+    this.tableLoader = tableLoader;
+    this.equalityFieldColumns = equalityFieldColumns;
+    this.uidPrefix = uidPrefix == null ? "" : uidPrefix;
+    this.snapshotProperties = snapshotProperties;
+
+    Preconditions.checkNotNull(tableLoader, "Table loader shouldn't be null");
+
+    if (newTable == null) {
+      tableLoader.open();
+      try (TableLoader loader = tableLoader) {
+        this.table = loader.loadTable();
+      } catch (IOException e) {
+        throw new UncheckedIOException("Failed to load iceberg table from table loader: " + tableLoader, e);
+      }
+    } else {
+      this.table = newTable;
+    }
+    FlinkWriteConf flinkWriteConf = new FlinkWriteConf(table, writeOptions, readableConfig);
+
+    this.equalityFieldIds = checkAndGetEqualityFieldIds(table, equalityFieldColumns);
+
+    this.flinkRowType = toFlinkRowType(table.schema(), tableSchema);
+
+    // Validate the equality fields and partition fields if we enable the upsert mode.
+    if (flinkWriteConf.upsertMode()) {
+      Preconditions.checkState(
+          !flinkWriteConf.overwriteMode(),
+          "OVERWRITE mode shouldn't be enable when configuring to use UPSERT data stream.");
+      Preconditions.checkState(
+          !equalityFieldIds.isEmpty(),
+          "Equality field columns shouldn't be empty when configuring to use UPSERT data stream.");
+      if (!table.spec().isUnpartitioned()) {
+        for (PartitionField partitionField : table.spec().fields()) {
+          Preconditions.checkState(equalityFieldIds.contains(partitionField.sourceId()),
+              "In UPSERT mode, partition field '%s' should be included in equality fields: '%s'",
+              partitionField, equalityFieldColumns);
+        }
+      }
+    }
+
+    this.upsertMode = flinkWriteConf.upsertMode();
+    this.overwrite = flinkWriteConf.overwriteMode();
+    this.distributionMode = flinkWriteConf.distributionMode();
+    this.workerPoolSize = flinkWriteConf.workerPoolSize();
+    this.targetDataFileSize = flinkWriteConf.targetDataFileSize();
+    this.dataFileFormat = flinkWriteConf.dataFileFormat();
+  }
+
+  @Override
+  public DataStream<RowData> addPreWriteTopology(DataStream<RowData> inputDataStream) {
+    return distributeDataStream(inputDataStream, equalityFieldIds, table.spec(),
+        table.schema(), flinkRowType, distributionMode, equalityFieldColumns);
+  }
+
+  @Override
+  public IcebergStreamWriter createWriter(InitContext context) {
+    return restoreWriter(context, null);
+  }
+
+  @Override
+  public IcebergStreamWriter restoreWriter(InitContext context,
+                                           Collection<IcebergStreamWriterState> recoveredState) {
+    StreamingRuntimeContext runtimeContext = runtimeContextHidden(context);
```

Comment:
This seems quite ugly. Are we reasonably sure that we have the correct objects here?

Reply:
Yes, I agree with you. Let me see if there's a better way.